src/pybind/mgr/dashboard/services/orchestrator.py

```diff
@@ -1,38 +1,101 @@
+    def status(self):
         try:
-            status, desc = super(OrchClient, self).available()
+            status, desc = super(OrchestratorAPI, self).available()
             logger.info("[ORCH] is orchestrator available: %s, %s", status, desc)
-            return status
+            return dict(available=status, description=desc)
         except (RuntimeError, OrchestratorError, ImportError):
-            return False
+            return dict(available=False,
+                        description='Orchestrator is unavailable for unknown reason')
+
+    def orchestrator_wait(self, completions):
+        return self._orchestrator_wait(completions)
+
+
+def wait_api_result(method):
+    @wraps(method)
+    def inner(self, *args, **kwargs):
+        completion = method(self, *args, **kwargs)
+        self.api.orchestrator_wait([completion])
+        raise_if_exception(completion)
+        return completion.result
+    return inner
+
+
+class ResourceManager(object):
+    def __init__(self, api):
+        self.api = api
+
+
+class HostManger(ResourceManager):
+
+    @wait_api_result
+    def list(self):
+        return self.api.get_hosts()
+
+    def get(self, hostname):
+        hosts = [host for host in self.list() if host.name == hostname]
+        return hosts[0] if hosts else None
 
-    def reload_service(self, service_type, service_ids):
+    @wait_api_result
+    def add(self, hostname):
+        return self.api.add_host(hostname)
+
+    @wait_api_result
+    def remove(self, hostname):
+        return self.api.remove_host(hostname)
+
+
+class InventoryManager(ResourceManager):
+
+    @wait_api_result
+    def list(self, hosts=None, refresh=False):
+        node_filter = InventoryFilter(nodes=hosts) if hosts else None
+        return self.api.get_inventory(node_filter=node_filter, refresh=refresh)
+
+
+class ServiceManager(ResourceManager):
+
+    @wait_api_result
+    def list(self, service_type=None, service_id=None, node_name=None):
+        return self.api.describe_service(service_type, service_id, node_name)
+
+    def reload(self, service_type, service_ids):
         if not isinstance(service_ids, list):
             service_ids = [service_ids]
 
-        completion_list = [self.service_action('reload', service_type,
-                                               service_name, service_id)
+        completion_list = [self.api.service_action('reload', service_type,
+                                                   service_name, service_id)
                            for service_name, service_id in service_ids]
-        self._orchestrator_wait(completion_list)
+        self.api.orchestrator_wait(completion_list)
         for c in completion_list:
             raise_if_exception(c)
+
+
+class OrchClient(object):
+    def __init__(self):
+        self.api = OrchestratorAPI()
+
+        self.hosts = HostManger(self.api)
+        self.inventory = InventoryManager(self.api)
+        self.services = ServiceManager(self.api)
```

Comment:
just fyi, The orchestrator interface aka `OrchestratorMixin` has no state and will never have. (well except that it needs to know the `mgr`.

Reply:
If you have a better way of giving `OrchestratorClientMixin` access to the dashboard's `mgr`, feel free to change it.

Reply:
> In that case I don't think it's needed to instantiate it, don't you think? Maybe it could be initialized once (singleton-like) or by a class method (`OrchClient.init()`) 

Yes. I'll refine this.

>If you have a better way of giving OrchestratorClientMixin access to the dashboard's mgr, feel free to change it.

My intention is to wrap the class that extends `OrchestratorClientMixin` in the client. Because I tend to implement methods that have similar names with methods in `OrchestratorClientMixin` (which extends `Orchestrator`). For example, `get_hosts`, `add_host`, and `remove_host`. A client that extends `OrchestratorClientMixin` make s me feel like I'm writing an orchestrator backend.

Reply:
The class is initialized once now, client code use `get_client()` to retrieve the instance. Please see https://github.com/ceph/ceph/pull/29127/commits/5062a39030f513efa5e36e13c5e69a5a9bb4aa4a

Reply:
Thanks!